### PR TITLE
Editgear

### PR DESCRIPTION
**Pull request by EllieLilithFox**
change EditGearForm to functional component and add tests, tweaks to Input component and tests. Write tests for mapStateToProps for EditGearForm and Form components.

### CODE REVIEW

File: gearcatalog/frontend/src/components/accounts/__tests__/Login.test.tsx [hunk 1]
```diff
@@ -9,7 +9,7 @@ jest.mock('react-router-dom', () => ({
   useHistory: () => ({
     push: mockHistoryPush,
   }),
-  Link: 'Link',
+  Link: 'div',
```
Comment (NatalieWolfe) [inline gearcatalog/frontend/src/components/accounts/__tests__/Login.test.tsx:12]: What is this change for?

File: gearcatalog/frontend/src/components/gear/EditGearForm.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  const [gearUpdated, setGearUpdated] = useState(false);
+  const [name, setName] = useState(selectedGear.name);
+  const [desc, setDesc] = useState(selectedGear.desc);
+  const [brand, setBrand] = useState(selectedGear.brand);
+  const [weight_grams, setWeight] = useState(selectedGear.weight_grams);
+  const [length_mm, setLength] = useState(selectedGear.length_mm);
+  const [width_mm, setWidth] = useState(selectedGear.width_mm);
+  const [depth_mm, setDepth] = useState(selectedGear.depth_mm);
+  const { value:locking, bind:bindLocking, reset:resetLocking } = 
+  useInput(selectedGear.locking);
```
Comment (NatalieWolfe) [inline gearcatalog/frontend/src/components/gear/EditGearForm.tsx]: Is this the formatting generated by your linter? It seems a bit odd to have wrapped this without indenting it.

File: gearcatalog/frontend/src/components/gear/types.ts [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
   desc: string,
   brand: string,
-  weight_grams: number,
-  length_mm: number,
-  width_mm: number,
-  depth_mm: number,
+  weight_grams: string,
+  length_mm: string,
+  width_mm: string,
+  depth_mm: string,
```
Comment (NatalieWolfe) [inline gearcatalog/frontend/src/components/gear/types.ts]: These should remain numbers. Your form should parse the input into numbers.

**Review by NatalieWolfe — CHANGES_REQUESTED** [review]

File: gearcatalog/frontend/src/components/accounts/__tests__/Login.test.tsx (hunk 1 above)
Comment (EllieLilithFox) [reply gearcatalog/frontend/src/components/accounts/__tests__/Login.test.tsx:12]: When I was running the tests it was giving me an `console.error` stating that react elements need to be pascal case. So I just changed the mock to a built in element.

File: gearcatalog/frontend/src/components/gear/EditGearForm.tsx (hunk 2 above)
Comment (EllieLilithFox) [reply gearcatalog/frontend/src/components/gear/EditGearForm.tsx]: No, I was just unsure if I needed to indent lines like that. I will indent it.

File: gearcatalog/frontend/src/components/common/Input.tsx [hunk 4]
```diff
@@ -1,10 +1,10 @@
-import React from "react";
+import React, { SetStateAction } from "react";
 
 type InputProps = {
   name: string,
   val: string | number,
   type: string,
-  setVal: (val: string) => void,
+  setVal: (val: SetStateAction<string> | SetStateAction<number>) => void,
```
Comment (NatalieWolfe) [inline gearcatalog/frontend/src/components/common/Input.tsx:7]: Check out generics for this. I believe this struct could be reshaped like this:

```ts
type InputProps<ValueType> = {
  name: string,
  val: ValueType,
  setVal: (val: SetStateAction<ValueType>) => void,
  ...
};
```

Then the input class would also be generisized in the same way.

```ts
const Input = <ValueType,>(props: InputProps<ValueType>): JSX.Element => {
  ...
};
```

Note that it is `<ValueType,>`, with a trailing comma, for the function. This is to break out of JSX's element parsing.

The bonus in doing it this way is it ensures correct typing between the `val` and `setVal` properties.

**Review by NatalieWolfe — CHANGES_REQUESTED** [review]

**Pull request merged**